### PR TITLE
[WIP] Add Attention is All you need transformer and Translation example

### LINKED ISSUE #148: Translation example
It'd be great to have a neural machine translation model incorporated into our models repository.

### PR DESCRIPTION
**Pull request by acarrera94**
This pull request adds the transformer from "attention is all you need"/"the annotated transformer" into the models folder. And adds a translation example for WMT'14 English-German data. 
I'd love to hear any feedback, and any help would be great!

This would close #148

TODO: 
- [x] implement transformer
- [x] add dataset
- [x] working training loop
- [x] clean up comments
- [x] split dataset into eval and add eval into training loop
- [ ] add testing dataset and test loop

### CODE REVIEW

**Comment by acarrera94** [comment]
> @googlebot I signed it!

**Comment by brettkoonce** [comment]
SUGOI DESU NE

File: Models/Translation/Attention.swift [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+//  TranslationTransformer
+//
+//  Created by Andre Carrera on 2/11/20.
+//  Copyright © 2020 Lambdo. All rights reserved.
+//
+
+import TensorFlow
+
+///// Input to an attention layer.
+//public struct AttentionInput<Scalar: TensorFlowFloatingPoint>: Differentiable {
```
Comment (brettkoonce) [inline Models/Translation/Attention.swift]: might clean this out if it's not in use

File: Examples/Transformer-Translation/main.swift [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+//
+
+
+import TensorFlow
+import TranslationModels
+import Foundation
+
+struct WMTTranslationTask {
+    // https://nlp.stanford.edu/projects/nmt/
+    // WMT'14 English-German data
```
Comment (brettkoonce) [inline Examples/Transformer-Translation/main.swift]: WMT == 💪!

**Comment by texasmichelle** [comment]
Thank you for putting this together! A translation example would indeed be a fantastic addition to the repo.

Between GPT-2 and BERT, we're trying to get a handle on the code duplication among transformer-based models. We'd like to reuse code among common concepts, which a translation example could also utilize. If you're up for it, we could use some help with #436 before compounding the problem by adding more transformer code.

Breaking this up into smaller PRs would also be helpful, such as the standalone dataset or at least leveraging existing components (because how many separate implementations of multi-head attention do we really need? Apropos multi-head attention, it appears you could easily use existing libraries since the code looks nearly identical, but your name is listed as author in the header so maybe there are some major changes).

**Comment by marcrasi** [comment]
Hi, there hasn't been activity on this PR for a while. We're going to close it to keep our list of open PRs small. Of course, feel free to reopen any time if you have time to come back and work on this!

**Pull request closed without merging**